### PR TITLE
Network integration tests

### PR DESCRIPTION
**Pull request by BedrockSquirrel**
### Why is this change needed?

- Adds a tool that can be used for testing network situations against a local dev network in-memory or one of our existing testnets

### What changes were made as part of this PR:

It adds a few new tools to play with and iterate on. The abstractions probably need improving but hopefully a useful first step.

/devnetwork **(a package for running a test network and orchestrating situations with its nodes)**
- DevNetwork interface represents a network that can be manipulated by tests
    * inherits the NetworkConnector interface so it can be used in network-tests
    * exposes `Start()`, `Stop()`, Getters for specific NodeOperators, etc.
- DefaultDevNetwork implements it with a bunch of configurable options
	
/networktest **(a package for creating tests that can run against a network - either local dev network or a testnet)**
- NetworkConnector interface represents a network that can be connected to by tests
    * exposes rpc address getters for nodes, exposes a faucet allocation function so tests can fund their sim accounts
    * DevNetworks are implementations of NetworkConnector, but we also have a TestnetConnector implementation using http faucet to allow tests to run against diff environments
- load test library (currently called `traffic` but don't love that name)
    * provides a mechanism to have X simulated users start running in separate go routines, transacting with the NetworkConnector
    * aims to be flexible so those simulated users could be doing more complex things or different things or randomised things
- NetworkTest interface represents a runnable test that requires a NetworkConnector to run against. NetworkTests may run other NetworkTests (e.g. a FullHealthTest might run half a dozen different functional tests)
- DevNetworkScenarioTest represents a runnable test that requires a DevNetwork to run against. These tests are used to reproduce scenarios around stopping/starting/breaking nodes etc.
A DevNetworkScenarioTest may run other DevNetworkScenarioTests as well as NetworkTests.

/simwallet
- this integration test helper bundles a user's private keys with an a client and provides high level functionality (similar to metamask or other user-facing software/hardware wallet)

Note: this PR also touches the db code to make the sqlite temp DB setup public for re-use from the dev network code.

### Definition of done

- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated 
- [ ] End-to-end tests run if required (see below)

### End-to-end tests

Running the end-to-end tests in the [obscuro-test repo](https://github.com/obscuronet/obscuro-test) is currently at the 
discretion of the developer prior to merging a PR. The intention is not to slow down development by having the longer 
running end-to-end tests as a PR gate run on each branch commit etc. To manually trigger a run pre-merge;

- Go to [run_local_tests](https://github.com/obscuronet/obscuro-test/actions/workflows/run_local_tests.yml)
- Click the "Run workflow" button
- Use the main branch of obscuro-test 
- Enter the name of the PR branch for go-obscuro
- Run the workflow

The run takes approximately 20 minutes and any failure will be reported in the workflow output. Should the tests fail 
the docker container output for the local testnet and all test artifacts will be added to the run and can be downloaded
for debugging and analysis, with a retention period of 2 days. 

Note that every PR merge to main will also trigger a run of post-merge tests, so even if you do not manually trigger 
pre-merge, tests will be run post-merge. The intention being to catch any issues fast on main to allow for quick 
resolution. The post-merge test output can be seen at 
[run_merge_tests](https://github.com/obscuronet/obscuro-test/actions/workflows/run_merge_tests.yml) and will show both 
the PR number and author in the list of runs.

### CODE REVIEW

File: integration/simulation/devnetwork/sim_l1.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package devnetwork
+
+import (
+	"github.com/obscuronet/go-obscuro/go/ethadapter"
+	"github.com/obscuronet/go-obscuro/integration/gethnetwork"
+)
+
+type L1Network struct {
+	clients     []ethadapter.EthClient
+	gethNetwork *gethnetwork.GethNetwork
```
Comment (tudor-malene) [inline integration/simulation/devnetwork/sim_l1.go:10]: Does this mean this setup doesn't support the fully in-memory mocked l1 nodes?

File: integration/simulation/devnetwork/sim_node.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+// InMemNodeOperator represents an Obscuro node playing a role in a DevSimulation
+//
+// Anything a node operator could do, that we need to simulate belongs here, for example:
+// * start/stop/reset the host/enclave components of the node
+// * provide convenient access to RPC clients for the node
+// * it might even provide access to things like the database files for the node if needed
+//
+// Note: InMemNodeOperator will panic when things go wrong, we want to fail fast in sims and avoid verbose error handling in usage
+type InMemNodeOperator struct {
```
Comment (tudor-malene) [inline integration/simulation/devnetwork/sim_node.go:35]: would be useful to have the same concept for the l1 nodes as well. 
Stopping the l1 node, etc, is a useful thing to test as well

File: integration/networktest/network_test_types.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package networktest
```
Comment (tudor-malene) [inline integration/networktest/network_test_types.go:1]: to be consistent, this should be called "interfaces.go"

File: integration/networktest/traffic/traffic.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package traffic
```
Comment (tudor-malene) [inline integration/networktest/traffic/traffic.go:1]: I think this needs a slightly better name and a readme.

File: integration/networktest/traffic/native_token.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+const (
+	LowVolume        = 2  // tx per sec
+	HighVolume       = 20 // tx per sec
+	_defaultNumUsers = 3
+)
+
+var _defaultTransferAmt = big.NewInt(10000000000)
+
+func NewNativeFundsSimRunner(transactionsPerSecond int) *nativeFundsSimRunner {
```
Comment (tudor-malene) [inline integration/networktest/traffic/native_token.go:30]: this should return `traffic.Runner`

File: integration/networktest/traffic/traffic.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"testing"
+	"time"
+
+	"github.com/obscuronet/go-obscuro/integration/networktest"
+)
+
+type Runner interface {
```
Comment (tudor-malene) [inline integration/networktest/traffic/traffic.go:10]: this needs comments.

Comment (tudor-malene) [reply integration/networktest/traffic/traffic.go:10]: shouldn't this be in  "interfaces.go" as well?
It is one of the main abstractions

File: integration/networktest/scenarios/restart_validator_enclave.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+	"github.com/obscuronet/go-obscuro/integration/networktest"
+
+	"github.com/obscuronet/go-obscuro/integration/networktest/functional"
+	"github.com/obscuronet/go-obscuro/integration/networktest/traffic"
+)
+
+type restartValidatorEnclaveScenarioTest struct{}
+
+func (r *restartValidatorEnclaveScenarioTest) Run(t *testing.T, network networktest.DevNetwork) error {
```
Comment (tudor-malene) [inline integration/networktest/scenarios/restart_validator_enclave.go:16]: for some reason this doesn't implement the `NetworkTest` interface.

since all seems to be compiling fine, it could mean that something is not 100% right

File: integration/networktest/network_test_types.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+type DevNetwork interface {
+	NetworkConnector
+	Start()
+	GetSequencerNode() NodeOperator
+	GetValidatorNode(idx int) NodeOperator
+	CleanUp()
+}
+
+// ScenarioTest defines a test that can be run against a DevNetwork, e.g. to explore failover scenarios, restarts etc.
+type ScenarioTest interface {
```
Comment (tudor-malene) [inline integration/networktest/network_test_types.go:38]: I'm not convinced about these ScenarioTest and NetworkTest abstractions.

The naming is definitely not right. 
"NetworkTest" - means you don't control nodes
"ScenarioTest" - means you control nodes

The point of these interfaces is to be able to create some batteries of tests that run in a sequence?

**Review by tudor-malene — CHANGES_REQUESTED** [review]
Thanks Matt.

I think this is pretty good and the abstractions are mostly right.
The biggest comment is around creating an operator abstractions for the L1 nodes, and to think about how we can integrate the MockEthereumNode thing

File: integration/simwallet/simwallet.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+const (
+	_maxReceiptWaitTime  = 30 * time.Second
+	_receiptPollInterval = 1 * time.Second
+)
+
+// SimWallet implements wallet.Wallet so it can be used for the legacy Wallet code.
+// But it aims to provide a wider range of functionality, akin to the software and hardware wallets that users interact with.
+// Note: SimWallet is **not** thread-safe for a single wallet (creates nonce conflicts etc.)
+type SimWallet struct {
```
[the remark this answers is not in the document]
Comment (BedrockSquirrel) [reply integration/simwallet/simwallet.go:29]: Yeah maybe not. I'll change to `UserWallet` unless you can think of a better name?

File: integration/simulation/devnetwork/sim_l1.go (hunk 1 above)
Comment (BedrockSquirrel) [reply integration/simulation/devnetwork/sim_l1.go:10]: Yeah v good point, I hadn't thought about it, just considered that the in-mem sim would live on its current form but better to make this thing flexible so it can be used to refactor them both.

File: integration/networktest/scenarios/restart_validator_enclave.go (hunk 7 above)
Comment (BedrockSquirrel) [reply integration/networktest/scenarios/restart_validator_enclave.go:16]: Yeah the ScenarioTest thing is probably too confusing and unhelpful. This is a scenario test but not a `NetworkTest`, because the NetworkConnector interface is just for transacting against it doesn't allow you control the nodes.

I.e. ScenarioTest is a NetworkTest  but NetworkTest is **not** a ScenarioTest. Maybe ScenarioTest should have been called NodeOrchestrationTest or something.

File: integration/networktest/traffic/traffic.go (hunk 4 above)
Comment (BedrockSquirrel) [reply integration/networktest/traffic/traffic.go:1]: Added a readme but can't really think of a better name. Some ideas: `load`, `simulation`, `useractivity`, `simtraffic`, `usertraffic`?

File: integration/networktest/functional/native_token.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
+package functional
```
Comment (tudor-malene) [inline integration/networktest/functional/native_token.go:1]: I still find the abstractions a bit confusing. I think there's some subtle overlap in responsibilities, making it a bit hard to reason about them.

I'll try to outline my understanding.

- `runner`: Entry point, where you setup and wire together the other components
- `traffic`: Something that just generates traffic against a network, but does not mess with the network.
- `functional` - I can't think of any meaningful difference from the "traffic" stuff.
- `scenarios` - where you can mess with the nodes.


Would something like this be possible?
In the entry point do the following:

1. set up the network you want to use. (I think the abstractions for this bit are right)

2. set up the network failure profile ( most of the the stuff that's in scenarios right now)
(this could provide some hooks: `onNodeStopped`, `onNodeStarted`). This only works against the local networks

3. set up the functional stuff (stuff that is in `traffic` and `functional`). 

4. Wire them together.

I'm thinking that to keep things decoupled there could be a standard lifecycle channel exposed where the network and the network failure and maybe something else could push events.

This channel is then injected everywhere so components can make decisions based on what's going on in the environment

**Review by tudor-malene — CHANGES_REQUESTED** [review]
Ended up leaving a more substantial comment.
Probably best to brainstorm on a call.

**Review by StefanIliev545 — COMMENTED** [review]
Amazing work @BedrockSquirrel! I need to go through the PR a couple more times to fully wrap my head around everything, but it definitely feels like the future of our testing code base.

I might have deducted inaccurately how the traffic bit works, but from the looks of it everything necessary is already there to create a test/scenario where the runner launches a separate process which will actually run the tests and then have a custom verifier that checks the result/logs of the process. Furthermore on top of the custom logic maybe run some generic verifiers that ensure the agents haven't managed to break protocol assumptions, examples being all submitted rollups are chained,  nodes survived & synchronized, etc. 

What I'm planning to do (probably obvious by now :D) is to add a test with a runner that simply execs "npx hardhat obscuro:test" with injected network environment created using the  properties from the network connector. For this scenario however the API might be a bit clunky in some spots. Note that by no means should you look into addressing them for this PR, I'll just list them in case you end up wrenching relevant code bits as it might help you:
- Ideally I'd want faucet allocation/prefunding to be defined on the test level
- I see contract deployments are still with a todo - in the hardhat case it would be great if this behaviour is overrideable on the test level. Code example of what I'm thinking: ```...env.LocalDevNetwork().WithPredeployer(callback).WithPostdeployer(callback)```
Where lets say the predeployer callback returns a structure containing required deployments and arbitrary data which is fed in the postdeployer. The predeployer/postdeployer is coming out of the new CI deployment approach. Previously it was  `deploy L1 -> run obscuro -> deploy L2` and now it is `deploy L1 Required -> start obscuro -> deploy L1 & L2` so in the example pre-deploy -> deploy & prepare only required system contracts to start node; post-deploy - Everything else required by the test runner(s) 
- Provisioning the wallet extension would be a bit weird. Most go tests use the authenticated rpc clients directly, but those can't be reused with hardhat. Now there already is a task that runs a wallet extension inside hardhat that we use in the CI when deploying, but reusing this for testing might be clunky since the wallet extension will live in a separate process and be hard to debug. Ideally when defining the environment, regardless of the type I'd like to be able to do `env.SomeNetwork().WithWalletExtensions(numToStart)` and have the go environment creator provision them and pick ports while on the scenario side they would be accessible in the network connector ( for deployments and traffic runners). In theory you can do something close to `go walletext.main/instantiate(ports, success sig channel)` the required number of times and in result we end up with a very close to a real env simulation. The transactions are made using the JS web3 libraries entirely process independently, while the entire protocol is in the same process space and debuggable end to end easily. 

Additionally there are some cool (IMO) features we can (maybe) add in the future:
- Provide the genesis (both L1 & L2) on test level. Allows to bypass ordinary verification and bootstrap networks in broken states. Can also be used to model attack scenarios @tudor-malene  described in the upgradability PoS research. Might be a very involved process to craft the scenarios initially, but I reckon with time we can evolve some tooling to easily model and iterate on attack scenarios. I imagine with the fast finality work you might have even better ideas on how to approach creating a real environment with the starting point being a mocked state definition
- In R3 we threw up together a tool that allowed to programatically define environments and it would orchestrate the processes and automatically proxy all their network access. I don't remember what we used to create the fake proxying, but this looks very close https://blog.worldline.tech/2020/11/26/simulate-datacenters-latency-with-linux.html
The proxying was basically needed so that we could introduce network issues on the fly - insane latency, packet loss, net up down and other common issues in prod. I think it would be cool if we had the same capability in the network envs here. Essentially all the endpoints each node needs get individually proxied. I imagine we'd find some issues and a ton of places where we need to optimize the code just by running a simulation with 40-50ms latency and 15-25% packet loss.
- Being able to send signals through the network connector would also be amazing. Think SIGSTOP/SIGCONT to forcefully desynchronize nodes, sending a ton of transactions and hitting SIGTERM to test sequencer HA reliability and even doing brutal random SIGKILLs to probe for any weird data corruption holes we might have missed.

File: integration/networktest/traffic/traffictest.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	numSimUsers   int
+	actionsPerSec int
+}
+
+func (t *durationTrafficTest) Run(network networktest.NetworkConnector) error {
+	err := t.runner.Start(network)
+	if err != nil {
+		return err
+	}
+	time.Sleep(t.duration)
```
[the remark this answers is not in the document]
Comment (BedrockSquirrel) [reply integration/networktest/traffic/traffictest.go:44]: I think there are some sleeps in this PR  that I hacked in to be cautious that we should get rid of but his was the only one I thought was definitely necessary haha.

The test (simulated user traffic) is supposed to run for a certain duration before we stop it and verify the results, so this sleep happens on a separate thread to the runner and then stops it.

Maybe there's a more idiomatic way I should have done this or I should add a comment to clarify that runner starts its own goroutines?

File: integration/networktest/traffic/traffictest.go [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+	t.runner.Stop()
+	return t.verifier(t.runner.RunData(), network)
+}
+
+func (t *durationTrafficTest) Name() string {
+	return fmt.Sprintf("traffic-%s", t.runner.Name())
+}
+
+// WithVerifier is a construction Option for the DurationTest that allows you to override the ver
+func WithVerifiers(verifiers []RunVerifier) TestOpt {
```
[the remark this answers is not in the document]
Comment (BedrockSquirrel) [reply integration/networktest/traffic/traffictest.go:54]: Oh yeah bugger, I must have started writing this and then never finished. I think I'd meant to change the verifer field on the test to be a slice of verifiers and this was going to be a way to add them to a test. I'll get rid of it for now.

**Comment by BedrockSquirrel** [comment]
Thanks for the comments @StefanIliev545, generally agree with everything you said. A few thoughts as a I read through:

- yeah definitely intended the traffic tests as having separate running of tests and verification. Might not have got the abstractions perfect yet but we can tweak, but the general thinking was that:
   - verifer has access to the records of the test actions and the network to query the state and make further tests
   - want to be able to apply multiple verifiers, I've only made one that checks actions were successful so far but had planned to add others with correctness of balances, looking out for slowness (either on RPC response times or on transactions getting minted), and general post-test health check. And then I imagine there will be a bunch of others depending on what people are testing for.

- love the idea of the environment hooks to override the network setup/lifecycle behaviour (`WithPredeployer()` etc.), that will keep things flexible. As long as everything has convenient default behaviours (so simple tests are easy to write), I think we should make this thing as configurable/overridable as possible

- I agree we should have a convenient way to test with wallet extensions without starting separate processes. We should keep a clear distinction between the network environment and the test that's running though imo (like network env may survive to be used against multiple tests but I imagine a test will setup what it needs and clean up after itself). 

    I think I've made that a bit unclear with the `environment` interface. I see `environment` as just setting up the network side of things and not knowing anything about the tests, i.e. the sort of things that tests expect to be there in the network but that might last for multiple tests

    So I'd imagine any setting up of wallet extensions would happen as a prestep on the test (at the same time that user wallets etc. are being created and funded), rather than as a property of the environment. I.e. I imagine using wallet extensions might look more like this, where the sim users can be setup to run with wallet extensions instead of auth clients and the test deals with starting them up and tearing them down:

    `Run(t, env.LocalDevNetwork(), traffic.DurationTest(traffic.NativeFundsTransfers(), 30*time.Second, traffic.WithWalletExtensionUsers(true)))`

    Does that make sense or am I getting confused? Be good to chat about how you see it working with the hardhat stuff too, I've got a bit of a blindspot there and need to get up to speed :)

- And yeah love the extra features you mentioned, agree they should be fairly straightforward to implement using the NodeOperator interface as the gateway to screwing around with nodes. Should be able to do things like having the NodeOperator simulate a node losing its L1 connectivity or simulating a malicious host sending bad stuff to the enclave too and making sure the rest of the network reacts appropriately.

**Comment by StefanIliev545** [comment]
@BedrockSquirrel 
`Run(t, env.LocalDevNetwork(), traffic.DurationTest(traffic.NativeFundsTransfers(), 30*time.Second, traffic.WithWalletExtensionUsers(true)))`
This makes sense yes. I just need the rpc url coming from the network connector to just point to a running wallet extension and hardhat will work.

**Comment by StefanIliev545** [comment]
@BedrockSquirrel 

I don't think separating the execution from verification and bridging the gap with actions is a good idea. On the execution side we end up with what is essentially a state machine and on the verification side we end up with the state transitions, but to verify them you kind of have to rerun the state machine. Or use some arbitary threshold checks like the current sims do. I much prefer having asynchronous "flows" - I have balance X. so I must be able to transfer X. This succeeded so I expect the user i transferred to to be able to take this action successfully. If something is not right the flow signals a failure on the spot. 

With the actions verifications would be kind of weird - this user transfer, so this transaction receipt should be successful. Okay but did the receiving user had a balance AT this point in time as his current balance might have arrived from a different flow. Did the receiver execute the following action successfully. Essentially same as execution flow but separated and hard to follow. Runner and verifier might not really match.

Perhaps I'm not really understanding the direction you are taking. I can definitely see a need for generic post run verifiers that inspect the state of the protocol and ensure everything adds up, but the actions feel like a way to simply recreate the transaction tracker which isn't a very good verification tool in the first place since it never had much of a context given to it. Same will happen with the verifiers based on the actions unless they recreate the context. "There is this transaction". Okay but should this transaction succeed? Should it fail? With what reason? Is the resulting state okay? But the resulting state is a combination of a previous transaction and this transaction so what is the previous one?

**Comment by BedrockSquirrel** [comment]
> I just need the rpc url coming from the network connector to just point to a running wallet extension and hardhat will work.

I think maybe we're still disagreeing a bit here, probably because of my hardhat ignorance. NetworkConnector interface can't know about wallet extension RPC addresses. It's the users that have wallet extensions rather than the network. I.e. if the network RPC addresses were for a wallet extension then normal AuthClient tests wouldn't be able to use those RPC endpoints.

File: integration/networktest/runner.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer envCleanup()
+	fmt.Println("Preparing to run test:", test.Name())
+	err = test.Run(network)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fmt.Println("Test succeeded:", test.Name())
```
Comment (tudor-malene) [inline integration/networktest/runner.go:29]: the test might have failed.
I'd say a better message is: "Finished Test:"
And then the first would be: "Starting Test:|

File: integration/networktest/traffic/runner.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/obscuronet/go-obscuro/integration/networktest"
+)
+
+// Runner is used to simulate traffic against a durationTrafficTest network
```
Comment (tudor-malene) [inline integration/networktest/traffic/runner.go:11]: why only against a durationTrafficTest?

File: integration/networktest/traffic/runner.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+// When Start() is called then simulated traffic will begin against the provided NetworkConnector, until Stop() is called.
+// Then Verify() can be used to check the simulation worked as expected (it might check receipts, balances, response times etc.)
```
Comment (tudor-malene) [inline integration/networktest/traffic/runner.go:13]: there is no verify()  in this interface

File: integration/networktest/traffic/runner.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+type Action interface {
+	Name() string
+}
+
+const (
+	SendTransaction = "SendTransaction"
+)
+
+type ActionRecord interface {
```
Comment (tudor-malene) [inline integration/networktest/traffic/runner.go:38]: I'm a big fan of this abstraction.
I'm wondering if we could take this abstraction one step further, and get rid of some complexity.

For example the: `prepareSim` in the `randomTransactingSimRunner`.

A test itself could be a sequence of Actions performed in series of parallel.

- `UserActionItem` could contain the action logic to execute whatever it has to execute, and also to verify that it was executed successfully.

The "deploy" step in the prepareSim would be a normal Action.

-`OpsActionItem` - could be that restarting of the enclave

These actions could be wired to run in series or parallel. 
If in series, then the result of the previous execution should be available to the next

**Review by tudor-malene — COMMENTED** [review]
Left a comment about Actions

File: integration/networktest/env/dev_network.go [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+	"github.com/obscuronet/go-obscuro/integration/networktest"
+	"github.com/obscuronet/go-obscuro/integration/simulation/devnetwork"
+)
+
+type devNetworkEnv struct{}
+
+func (d *devNetworkEnv) Prepare() (networktest.NetworkConnector, func(), error) {
+	devNet := devnetwork.DefaultDevNetwork()
+	devNet.Start()
```
Comment (otherview) [inline integration/networktest/env/dev_network.go:14]: Likely that there is a swallowed error here

File: integration/networktest/env/dev_network.go [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+func awaitNodesAvailable(_ networktest.NetworkConnector) error { //nolint:unparam
+	// todo: create RPC clients for all the nodes and wait until their health checks pass
+
+	// for now we just sleep
+	time.Sleep(15 * time.Second)
+	return nil
+}
+
+func LocalDevNetwork() networktest.Environment {
```
Comment (otherview) [inline integration/networktest/env/dev_network.go:32]: This method is typically called NewDevNetwork() to make clear it's returning a new instance

File: integration/networktest/env/network_setup.go [hunk 19]
```diff
@@ -0,0 +1,32 @@
+package env
```
Comment (otherview) [inline integration/networktest/env/network_setup.go:1]: Removing the `testnetEnv` can simplify things. Does not seem like it's reused and if the`networktest.Environment` interface changes the `testnetEnv` and the underlying `testnetConnector` have to be changed.

Would prolly mean that the `network_setup.go` file would be removed and files could pattern off with `dev_network.go` and `testnet_network.go` ?

File: integration/networktest/env/testnet.go [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+func (t *testnetConnector) AllocateFaucetFunds(account gethcommon.Address) error {
+	req := map[string]string{"address": account.Hex()}
+	jsonPayload, _ := json.Marshal(req)
+	resp, err := http.Post(t.faucetHTTPAddress, "application/json", bytes.NewBuffer(jsonPayload)) //nolint:noctx
+	if err != nil {
+		return fmt.Errorf("unable to make http faucet request - %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
```
Comment (otherview) [inline integration/networktest/env/testnet.go:41]: would `resp.StatusCode != 200` be ok here ?

File: integration/networktest/tests/ci/full_network_sim_test.go [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+	"testing"
+	"time"
+
+	"github.com/obscuronet/go-obscuro/integration/networktest"
+	"github.com/obscuronet/go-obscuro/integration/networktest/env"
+	"github.com/obscuronet/go-obscuro/integration/networktest/traffic"
+)
+
+// TestFullNetworkSim reimplements the test case of our current "sim" tests (the ERC20 transaction injector)
+func TestFullNetworkSim(t *testing.T) {
```
Comment (otherview) [inline integration/networktest/tests/ci/full_network_sim_test.go:13]: Should this be removed until the sims are replaced ?

File: integration/simulation/devnetwork/mock_l1.go [hunk 22]
```diff
@@ -0,0 +1,54 @@
+package devnetwork
```
Comment (otherview) [inline integration/simulation/devnetwork/mock_l1.go:1]: This file should probably be removed

**Review by otherview — COMMENTED** [review]
Really good stuff. I think the scope as been strechted and also there are parts here with unused code, commented out stuff and so on.

This tooling is very much helpful but I'd reduce the number of things it's trying to achieve on this PR.

Getting a version of smoke tests for dev use in main - for example - would be amazing imo. I'd prefer reducing the scope of the PR and getting it merged ASAP than to try and get all the new features merged in this one PR.
Plus it would make it easier for someone to build on top of it :)

File: integration/simulation/devnetwork/mock_l1.go (hunk 22 above)
Comment (BedrockSquirrel) [reply integration/simulation/devnetwork/mock_l1.go:1]: whoops, thanks

File: integration/networktest/tests/ci/full_network_sim_test.go (hunk 21 above)
Comment (BedrockSquirrel) [reply integration/networktest/tests/ci/full_network_sim_test.go:13]: Yeah probably, I'll get rid of the erc20 placeholder too.

File: integration/networktest/env/network_setup.go (hunk 19 above)
Comment (BedrockSquirrel) [reply integration/networktest/env/network_setup.go:1]: It's definitely ugly under the hood but I think it's worth if for the interface personally. The environment abstraction makes it easy to change what you are running a test against. I think `env.LocalDevNetwork()` helps with that usage too but we can discuss.

File: integration/networktest/env/dev_network.go (hunk 17 above)
Comment (BedrockSquirrel) [reply integration/networktest/env/dev_network.go:14]: Not swallowed, but panicked. The dev network stuff panics if it fails during start-up  at the moment, since it's test code but maybe that should be changed.

File: integration/networktest/traffic/runner.go (hunk 14 above)
Comment (BedrockSquirrel) [reply integration/networktest/traffic/runner.go:11]: This is the traffic test runner, the durationTrafficTest is how it's implemented I guess. Just simulates activity for a duration then runs the configured verifiers against it. But I think maybe this stuff needs a rethink since the interface doesn't seem to be intuitive.

File: integration/networktest/env/dev_network.go (hunk 17 above)
Comment (otherview) [reply integration/networktest/env/dev_network.go:14]: Yep, worth surfacing all errors.

File: integration/networktest/env/network_setup.go (hunk 19 above)
Comment (otherview) [reply integration/networktest/env/network_setup.go:1]: Can still use the same approach and maintain the same methods `env.Testnet()`, `env.DevTestnet()` `env.LocalDevNetwork`.
It's the middle layer `testnetEnv` that's not really needed nor reused. It wraps a `testnetConnector`.
Both `Testnet` `DevTestnet` could just return a compatible `networktest.Environment` struct.
I might be missing something tho, chat when you have tick :) ?

Comment (BedrockSquirrel) [reply integration/networktest/env/network_setup.go:1]: "could just return a compatible networktest.Environment struct" is that not exactly what testnetEnv is? It's just a wrapper to make the testnet connector compatible with the Environment interface

Comment (otherview) [reply integration/networktest/env/network_setup.go:1]: Yeah exactly. I might be missing something, but why wrap instead of making testnet connector compatible with the  networktest.Environment  interface ? ( I think it would be adding the Prepare method ? )

**Comment by BedrockSquirrel** [comment]
Closing this pull request as there's been multiple major revisions and the comment chain is hard to follow. Continuing on fresh branch here: https://github.com/obscuronet/go-obscuro/pull/1039
